os/fs/smartfs/smartfs_utils.c

```diff
@@ -1452,11 +1388,78 @@ int smartfs_deleteentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s *en
 
 				sector = SMARTFS_NEXTSECTOR(header);
 			}
+		} 
+	}
+
+	if (inactive_entry == TRUE) {
+		/* OK, another entry exist, inactive current one */
+		readwrite.logsector = entry->dsector;
+		readwrite.offset = 0;
+		readwrite.count = fs->fs_llformat.availbytes;
+		readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
+		/* TODO We read again here to mark that entry as inactive, let's find a way to avoid read 2 times */
+		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+		if (ret < 0) {
+			fdbg("Error reading directory info at sector %d\n", entry->dsector);
+			goto errout;
+		}
+	
+		/* Mark this entry as inactive */
+	
+		direntry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[entry->doffset];
+#if CONFIG_SMARTFS_ERASEDSTATE == 0xFF
+#ifdef CONFIG_SMARTFS_ALIGNED_ACCESS
+		smartfs_wrle16(&direntry->flags, smartfs_rdle16(&direntry->flags) & ~SMARTFS_DIRENT_ACTIVE);
+#else
+		direntry->flags &= ~SMARTFS_DIRENT_ACTIVE;
+#endif
+#else							/* CONFIG_SMARTFS_ERASEDSTATE == 0xFF */
+#ifdef CONFIG_SMARTFS_ALIGNED_ACCESS
+		smartfs_wrle16(&direntry->flags, smartfs_rdle16(&direntry->flags) | SMARTFS_DIRENT_ACTIVE);
+#else
+		direntry->flags |= SMARTFS_DIRENT_ACTIVE;
+#endif
+#endif							/* CONFIG_SMARTFS_ERASEDSTATE == 0xFF */
+	
+		/* Write the updated flags back to the sector */
+		
+		readwrite.offset = entry->doffset;
+		readwrite.count = sizeof(uint16_t);
+		readwrite.buffer = (uint8_t *)&direntry->flags;
+		ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
+		if (ret < 0) {
+			fdbg("Error marking entry inactive at sector %d\n", entry->dsector);
+			goto errout;
 		}
 	}
+	/* Now Free Chained sector from target entry */
+	
+	nextsector = entry->firstsector;
+	header = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+	readwrite.offset = 0;
+	readwrite.count = sizeof(struct smartfs_chain_header_s);
+	readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
+	while (nextsector != SMARTFS_ERASEDSTATE_16BIT) {
+		/* Read the next sector into our buffer */
 
-	ret = OK;
+		sector = nextsector;
+		readwrite.logsector = sector;
+		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+		if (ret < 0) {
+			fdbg("Error reading sector %d\n", nextsector);
+			break;
+		}
 
+		/* Release this sector */
+
+		nextsector = SMARTFS_NEXTSECTOR(header);
+		ret = FS_IOCTL(fs, BIOC_FREESECT, sector);
+		if (ret < 0) {
+			fdbg("Error freeing sector %d\n", nextsector);
+			break;
+		}
+	}
+	
 errout:
 	return ret;
 }
```

Comment:
All functions are so long. Is it possible to separate the code for easy maintenance?

Reply:
I already know, not only this file long functions are spread. 
There are 2 reason why I keep current state,
One is if there are something big size duplicated code, I will make it function. But nothing here. Then we'd better reduce jump time, we are rtos.
Another one is sometimes we brought bug fix or something useful from our Mommy's repository, so as possible as I can, I prefer to keep originality of code.
Enough?

Reply:
ok. I agree